mobile/ios/ios-device.ts

```diff
@@ -320,11 +327,130 @@ export class IOSDevice implements Mobile.IIOSDevice {
+        notificationProxyClient.addObserver(appReadyForAttachMessage, () => {
+            this.$logger.info("Got ReadyForAttach");
+
+            var devicePort = 8080;
+            var localPort = 8080;
+
+            var localServer = net.createServer((localSocket) => {
+                this.$logger.info("Front-end client connected to localhost " + localPort);
+
+                var deviceSocket: any;
+
+                var buff = new Array();
+                localSocket.on('data', (data: NodeBuffer) => {
+                    if (deviceSocket) {
+                        deviceSocket.write(data);
+                    } else {
+                        buff.push(data);
+                    }
+                });
+                localSocket.on('end', () => {
+                    this.$logger.info('Localhost client disconnected!');
+                    process.exit(0);
+                });
+
+                deviceSocket = this.connectToPort(devicePort);
+                this.$logger.info("Connected localhost " + localPort + " to device " + devicePort);
+
+				buff.forEach((data) => {
+					deviceSocket.write(data);
+				});
+
+				deviceSocket.on('data', (data: NodeBuffer) => {
+					localSocket.write(data);
+				});
+
+				deviceSocket.on('end', (data: NodeBuffer) => {
+                    this.$logger.info("Device socket closed!");
+					process.exit(0);
+				});
+
+			});
+
+            localServer.listen(localPort, () => {
+                this.$logger.info("Opened localhost " + localPort);
+            });
+        });
+    }
+
+    private printHowToTerminate() {
+        this.$logger.info("\nSetting up debugger proxy...\n\nPress Ctrl + C to terminate!\n");
+    }
+
+    private postDebugNotification(packageName: string, notification: string, notificationProxyClient: iOSProxyServices.NotificationProxyClient): void {
+        var attachRequestMessage = packageName + ":NativeScript.Debug.AttachRequest";
+        notificationProxyClient.postNotification(attachRequestMessage);
+        this.$logger.info("Send " + notification);
+    }
+
+    private attachAtStartup(packageName: string) {
+        this.openSafariFrontEnd(packageName);
+
+        var _injector = this.$injector;
+        var notificationProxyClient: iOSProxyServices.NotificationProxyClient = _injector.resolve(iOSProxyServices.NotificationProxyClient, { device: this });
+
+        this.printHowToTerminate();
+
+        var appLaunchMessage = packageName + ":NativeScript.Debug.AppLaunching";
+        notificationProxyClient.addObserver(appLaunchMessage, () => {
+            this.$logger.info("Got AppLaunching");
+
+            this.attachWhenReady(packageName, notificationProxyClient);
+            this.postDebugNotification(packageName, "WaitForDebugger", notificationProxyClient);
+        });
+    }
+
+    private attachAtRuntime(packageName: string): void {
+        this.openSafariFrontEnd(packageName);
```

Comment:
This function is very similar to attachAtStartup, can we refactor the common parts?